### PR TITLE
fix: discard outputs when rustc inputs mutate

### PR DESCRIPTION
**Pull request by jdx**
## Summary

- distinguish compiler-input mutation from ordinary cache publication failures
- remove every modeled rustc output and return failure when inputs change during or immediately after compilation
- update learned-incremental state only after compiler inputs have been verified
- cover the race with a deterministic delayed-compiler integration test

This addresses the correctness failure reported in [discussion #325](https://github.com/jdx/mr-boxington/discussions/325): Cargo can no longer compile a dependent against metadata produced before a mid-compilation edit.

## Testing

- `cargo test -p mbx`
- `cargo clippy -p mbx --all-targets -- -D warnings`

---

> [!NOTE]
> **Medium Risk**
> Touches the core rustc wrapper path (timing of snapshots, output replay, and failure semantics) that Cargo depends on for pipeline ordering; behavior is heavily tested but wrong classification could either fail good builds or briefly expose stale artifacts.
> 
> **Overview**
> Fixes a race where Cargo could build dependents against **metadata from a compilation that started before a mid-build source edit**.
> 
> **Input snapshots** are taken **after** the scheduler admits the compile (not before), so edits while waiting for capacity are treated as part of the upcoming compile rather than as overlap during rustc.
> 
> After a successful rustc run, **publication/validation** runs before **stdout is replayed** to Cargo, so artifact notifications are not emitted for results that will be rejected.
> 
> When validation fails only because **compiler inputs actually changed** (`BypassReason::InputChanged` or overlap proven by pre-compile **content** identity), the wrapper **deletes modeled outputs** (artifacts + dep-info), logs an error, replays stderr only, and returns **failure**. Ordinary cache/publication errors still warn and leave rustc’s local outputs usable. **Timestamp-only** overlap during compile does **not** fail the build.
> 
> **Learned incremental** paths reuse manifest-matched discovered inputs for a lighter verify path; churn recording moves to **after** input checks. **Execution-only build scripts** get the same validation.
> 
> Tests cover classification/cleanup helpers and Unix e2e builds with a **delayed rustc** wrapper that edits sources after compile finishes but before the wrapper returns.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f1c9894bbe53ad69fb8aadaf37b34b30b9d810df. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved compilation validity checks to detect source changes made while builds are waiting for capacity.
  * Stale compiler outputs are now removed when compiler inputs change during compilation.
  * Compilation now reports modified inputs as errors instead of returning potentially stale results.
  * Unrelated publication or caching failures continue to produce warnings without deleting outputs.

* **Tests**
  * Added coverage for dependency and dependent-source changes during compilation.
  * Verified cleanup of metadata, library, and dependency-information outputs, including repeated cleanup attempts.

### CODE REVIEW

File: crates/mbx/src/rustc.rs [hunk 1]
```diff
@@ -536,6 +537,15 @@ pub(crate) fn compile(
...
             Ok(diagnostic) => current_diagnostic = diagnostic,
+            Err(error) if compiler_input_was_modified(&error) => {
+                if let Err(discard_error) = discard_compiler_outputs(&outputs) {
+                    eprintln!(
+                        "mbx[warning]: some invalid compiler outputs could not be removed: {discard_error:#}"
+                    );
+                }
+                eprintln!("mbx[error]: compilation result was discarded: {error:#}");
+                final_status = ExitCode::FAILURE;
+            }
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/rustc.rs:598]: Good catch. The required-input snapshot now happens after scheduler admission and immediately before `compilation_started`/rustc launch, so edits made while waiting for capacity are part of the upcoming valid compilation rather than classified as overlap. I also kept the deterministic mid-rustc mutation test passing with the corrected boundary.

**Comment by jdx** [comment]
Addressed the CodeRabbit review in 71d8ccf. Both alternate successful compiler paths now snapshot required inputs after scheduler admission, validate the compiler-reported inputs after rustc exits, discard modeled outputs on a detected mutation, and return failure. The execution-only build-script path also has a deterministic regression test that edits build.rs after rustc writes its output.

File: crates/mbx/src/rustc.rs [hunk 2]
```diff
@@ -488,77 +506,96 @@ pub(crate) fn compile(
...
+                    );
+                    Ok(compilation_action_diagnostic(&compilation, action).ok())
+                })()
+            };
+        match publication {
+            Ok(diagnostic) => current_diagnostic = diagnostic,
+            Err(error) if discard_modified_compiler_result(&outputs, &error) => {
+                compiler_input_invalid = true;
             }
+            Err(error) => eprintln!("mbx[warning]: result was not stored: {error:#}"),
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/rustc.rs:599]: Fixed in f1c9894. Rustc stdout is now withheld until input validation and publication complete on normal, verification, and execution-only build-script paths. If the result is rejected, only stderr diagnostics are replayed, so Cargo cannot observe an artifact notification for outputs that are being discarded. The full mbx test suite passes, including both deterministic mid-compilation mutation tests.

File: crates/mbx/src/rustc.rs [hunk 3]
```diff
@@ -567,7 +604,80 @@ pub(crate) fn compile(
...
+    if !compiler_input_was_modified(error) {
+        return false;
+    }
+    if let Err(discard_error) = discard_compiler_outputs(outputs) {
+        eprintln!(
+            "mbx[warning]: some invalid compiler outputs could not be removed: {discard_error:#}"
+        );
+    }
+    eprintln!("mbx[error]: compilation result was discarded: {error:#}");
+    true
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/rustc.rs:659]: Fixed in f1c9894. Fatal invalidation now requires either a verified content change or an InputModifiedDuringCompilation path backed by a pre-compilation identity with a change token. Timestamp-only overlap for dep-info-only inputs remains a nonfatal cache bypass, avoiding hard failures from NFS or clock skew. Added a focused regression test, and the full mbx test suite passes.
